samples/ServerSideBlazor.Server/Startup.cs

```diff
@@ -0,0 +1,37 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.DependencyInjection;
+using ServerSideBlazor.Client;
+
+namespace ServerSideBlazor.Server
+{
+    public class Startup
+    {
+        public void ConfigureServices(IServiceCollection services)
+        {
+            // Adds the Server-Side Blazor services, and those registered by the client startup.
+            services.AddServerSideBlazor<Client.Startup>();
+
+            // Since Blazor is running on the server, we can use an application service
+            // to read the forecast data.
+            services.AddSingleton<WeatherForecastService, DefaultWeatherForecastService>();
```

Comment:
This is really interesting. It seems like this addresses two different matters:

1. It's a pattern for providing some abstraction around data access (or other services) so that you could make a single Blazor app that runs either server-side *or* client-side, and inject a hosting-model-specific implementation during startup.
2. It's a way of getting around the project reference going in the Server->Client direction. In some ways it would feel more natural here if the reference was Client->Server so that the client app could directly just consume some classes from the server.

Point (1) sounds good. Not too many devs will need that level of flexibility, but it's good that it's there if needed.

Point (2) is a bit more contentious. I suspect at least in the early days we're going to get a lot of devs saying "why do I need to add these extra interfaces/ABCs just to consume services in my Blazor client app? it would be easier if it was just one project not two". Maybe in the long run it might push us towards making it possible to do it all in a single project, and somehow dealing with the compile-time complexity this causes for us.

In the short term, for concern (2), perhaps we would say "if it's a service you only need in your Blazor UI, then put it in the Blazor project" and "if you need the service both in the Blazor UI *and* to expose as a traditional JSON web API, then put it in a `.Shared` project and reference it from both `.Server` and `.Client`".

What do you think?

Reply:
@vertonghenb No that's not correct. There is only one service provider, so you only register things once. The reason we support having `Startup.cs` in both the server host project and the Blazor library project is so you can register services that are internal to both projects if you want (and generally encapsulate concerns in either project if you want).

Example: If you have a `UserRepository`, you'd register it just once, and it's available both to Blazor components and to (say) MVC actions.

Reply:
> What do you think?

I think that if this is the case the implementations would be different in client and server. For instance, I didn't create a client-side implementation of this repository, but if I had, I'm sure you can imagine how it would be different.